File: Day 3/Day 3.py
import re
import collections

with open('data.txt', 'r') as r:
    lines = r.read().splitlines()

total = 0
gears = collections.defaultdict(list)

def is_symbol(point):
    return (not point.isdigit() and point != '.')

def check_neighbors(lines, i, num, index):
    
    start = index - 1
    end = index + len(num)

    if start >= 0:
        if lines[i][start] == '*':
            gears[(i, start)].append(int(num))
        
        if is_symbol(lines[i][start]):
            return True

    if end < len(lines[i]):
        if lines[i][end] == '*':
            gears[(i, end)].append(int(num))
            
        if is_symbol(lines[i][end]):
            return True
    
    if start < 0:
        start = 0
    if end >= len(lines[i]):
        end = len(lines[i]) - 1
      
    neighboring_lines = []
    if i > 0:
        neighboring_lines.append((-1, lines[i-1]))
    if i < len(lines) - 1:
        neighboring_lines.append((1, lines[i+1]))

    for direction, line in neighboring_lines:
        for j, symbol in enumerate(line[start:end+1], start):
            if is_symbol(symbol):
                if symbol == "*":
                    gears[(i + direction, j)].append(int(num))
                return True
  
    return False
    
total = 0
gear_total = 0

for i, line in enumerate(lines):
    numbers = re.finditer("\d+", line)

    for match in numbers:
        num = match.group(0)
        if check_neighbors(lines, i, num, match.start()):
            total += int(num)

for gear in gears:
    if len(gears[gear]) == 2:
        x, y = gears[gear]
        gear_total += x * y

print("Part 1:", total)
print("Part 2:", gear_total)

File: Day 3/Day 3.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

var gears = make(map[[2]int][]int)

func isSymbol(point string) bool {
	return point != "." && !regexp.MustCompile(`^\d+$`).MatchString(point)
}

func checkNeighbors(lines []string, i int, num string, index int) bool {
	start := index - 1
	end := index + len(num)
	numInt, _ := strconv.Atoi(num)
	if start >= 0 {
		if lines[i][start:start+1] == "*" {
			gears[[2]int{i, start}] = append(gears[[2]int{i, start}], numInt)
		}

		if isSymbol(lines[i][start : start+1]) {
			return true
		}
	}

	if end < len(lines[i]) {
		if lines[i][end:end+1] == "*" {
			gears[[2]int{i, end}] = append(gears[[2]int{i, end}], numInt)
		}

		if isSymbol(lines[i][end : end+1]) {
			return true
		}
	}

	if start < 0 {
		start = 0
	}
	if end >= len(lines[i]) {
		end = len(lines[i]) - 1
	}

	neighboringLines := [][2]int{}
	if i > 0 {
		neighboringLines = append(neighboringLines, [2]int{-1, i - 1})
	}
	if i < len(lines)-1 {
		neighboringLines = append(neighboringLines, [2]int{1, i + 1})
	}

	for _, nl := range neighboringLines {
		direction, lineIndex := nl[0], nl[1]
		for j, symbol := range lines[lineIndex][start : end+1] {
			if isSymbol(string(symbol)) {
				if string(symbol) == "*" {
					gears[[2]int{i + direction, j + start}] = append(gears[[2]int{i + direction, j + start}], numInt)
				}
				return true
			}
		}
	}

	return false
}

func main() {
	data, err := ioutil.ReadFile("data.txt")
	lines := strings.Split(string(data), "\n")

	total := 0
	gearTotal := 0

	if err != nil {
		fmt.Println("File reading error", err)
		return
	}

	for i, line := range lines {

		re := regexp.MustCompile(`\d+`)
		numbers := re.FindAllString(line, -1)
		indices := re.FindAllStringIndex(line, -1)
		for j := 0; j <= len(numbers)-1; j++ {
			num := numbers[j]
			index := indices[j]
			if checkNeighbors(lines, i, num, index[0]) {
				num, _ := strconv.Atoi(num)
				total += num
			}
		}
	}

	for _, gear := range gears {
		if len(gear) == 2 {
			x, y := gear[0], gear[1]
			gearTotal += x * y
		}
	}

	fmt.Println("Part 1:", total)
	fmt.Println("Part 2:", gearTotal)
}
